### PR TITLE
Feat/elk: Add Elasticsearch/Kibana/Metricbeat

### PR DESCRIPTION
**Pull request by ghost**
This PR adds a demo setup for Elasticsearch/Kibana/Metricbeat stack.

### CODE REVIEW

File: docker/vault/config/statsd-telemetry.hcl [hunk 1]
```diff
@@ -0,0 +1,5 @@
+telemetry {
+  statsd_address = "host.docker.internal:8125"
```
Comment (ghost) [inline docker/vault/config/statsd-telemetry.hcl:2]: we can use the name `host.docker.internal` also in Podman deployments, since it is statically mapped in the `docker-compose` file

File: docker/metricbeat/statsd.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# Module: statsd
+# Docs: https://www.elastic.co/guide/en/beats/metricbeat/main/metricbeat-module-statsd.html
+
+- module: statsd
+  # listen on all interfaces for testing purposes
+  host: "0.0.0.0"
```
Comment (ghost) [inline docker/metricbeat/statsd.yml:6]: There is no actual risk by listening on all interfaces here since this is a test deployment. Production deployments could listen on the public interface only, of course

File: docker/metricbeat/statsd.yml [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+  port: "8125"
+  enabled: true
```
Comment (ghost) [inline docker/metricbeat/statsd.yml:8]: We have to deliberately enable the Statsd module here

File: docker/metricbeat/metricbeat.yml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+metricbeat.config.modules:
+  path: ${path.config}/modules.d/*.yml
+  reload.enabled: true
+
+processors:
+  - add_docker_metadata: ~
+
+output.elasticsearch:
+  hosts: ["${ELASTICSEARCH_HOSTS}"]
```
Comment (ghost) [inline docker/metricbeat/metricbeat.yml:9]: The list of hosts here is passed through the environment variable

File: docker/metricbeat/metricbeat.yml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+metricbeat.config.modules:
+  path: ${path.config}/modules.d/*.yml
+  reload.enabled: true
+
+processors:
+  - add_docker_metadata: ~
```
Comment (ghost) [inline docker/metricbeat/metricbeat.yml:6]: Not sure if this is actually relevant, but it was included in the default configuration file and I decided to keep it in here since we run it inside Docker

File: docker-compose.yaml [hunk 6]
```diff
@@ -72,6 +72,10 @@ services:
     ports:
       - 8200:8200
       - 8201:8201
+    extra_hosts:
+        # Makes this name resolve inside the container for Linux test hosts (used for Statsd Telemetry addresss)
+        # https://docs.docker.com/desktop/networking/#i-want-to-connect-from-a-container-to-a-service-on-the-host
+      - host.docker.internal:host-gateway
```
Comment (ghost) [inline docker-compose.yaml:78]: This is a static mapping to resolve the IP of the host inside the containers. The name will be used again in the Statsd telemetry stanza below.

File: .env_template [hunk 7]
```diff
@@ -3,3 +3,7 @@ export CONTAINER_DOMAIN=     # dns.podman or docker.internal
 export VAULT_LICENSE= # retrieve from https://www.hashicorp.com/products/vault/trial
 IPA_REALM=            # DNS.PODMAN or DOCKER.INTERNAL
 
+# Elasticsearch settings
+# https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html
+STACK_VERSION=8.4.1   # Version of Elastic products
+MEM_LIMIT=536870912   # 512M, increase or decrease based on the available host memory (in bytes)
```
Comment (ghost) [inline .env_template:9]: I tried to run it w/o any limits first (minimal config), but Elastic and Kibana decided to get all they can..

File: USAGE.md [hunk 8]
```diff
@@ -642,7 +647,7 @@ Also, add two additional Vault nodes, to build a complete 3-node cluster:
 +    cap_add:
 +      - IPC_LOCK
 +    networks:
-+      - docker.internal
++      - vault-playground
```
Comment (ghost) [inline USAGE.md:650]: this fix is actually not related to any topic of this PR here

File: USAGE.md [hunk 9]
```diff
@@ -843,3 +848,28 @@ EOF
...
+> https://docs.elastic.co/en/integrations/hashicorp_vault
+
+Export/Import: Go to "Stack Management > Saved Objects" to export/import the dashboard (https://www.elastic.co/guide/en/kibana/current/managing-saved-objects.html)
+
+Todo: Add and test API call here to import the `kibana-dashboard.ndjson` file automatically
+
+https://www.elastic.co/guide/en/kibana/current/saved-objects-api-import.html
+
+```bash
+curl -XPOST "http://127.0.0.1:5601/api/saved_objects/_import -H "kbn-xsrf: true" -d @kibana-dashboard.ndjson"
```
Comment (ghost) [inline USAGE.md]: This needs to be tested. We set `disable_hostname=true` in the listener to make the dashboard easy copy/pastable. Otherwise, the metrics have the domain name included which is different on each system.

File: docker-compose.yaml [hunk 10]
```diff
@@ -160,5 +164,50 @@ services:
...
+    hostname: "es.${CONTAINER_DOMAIN}"
+    networks:
+      - vault-playground
+    volumes:
+      - ./docker/es:/usr/share/elasticsearch/data:Z
+    ports:
+      - 9200:9200
+    environment:
+      - discovery.type=single-node
+      - xpack.security.enabled=false
```
Comment (ghost) [inline docker-compose.yaml:179]: we disable the Elasticsearch password authentication on the API here, since this is a playground (minimal config)

File: docker-compose.yaml [hunk 11]
```diff
@@ -160,5 +164,49 @@ services:
...
+    image: docker.elastic.co/beats/metricbeat:${STACK_VERSION}
+    container_name: metricbeat
+    hostname: "metricbeat.${CONTAINER_DOMAIN}"
+    networks:
+      - vault-playground
+    volumes:
+      - ./docker/metricbeat/metricbeat.yml:/usr/share/metricbeat/metricbeat.yml
+      - ./docker/metricbeat/statsd.yml:/usr/share/metricbeat/modules.d/statsd.yml
+    ports:
+      - 8125:8125/udp
```
Comment (ghost) [inline docker-compose.yaml:206]: - Metricbeat container listens on udp port 0.0.0.0:8125
- We map that port to the host system on port 8125 as well here
- The Vault server will send metrics to the IP of the Docker host (also port 8125)
- Metrics are then forwarded from Metricbeat to Elasticsearch

**Review by Xelef2000 — APPROVED** [review]
